File: Back-tracking/131-partition/go/131.go
package main

import "fmt"

var res [][]string
var path []string

func palindrome(s string) bool {
	l := 0
	r := len(s) - 1
	for l <= r {
		if s[l] != s[r] {
			return false
		}
		l++
		r--
	}
	return true
}

func backTracking(s string, cutPos int) {
	if cutPos == len(s) {
		tmp := make([]string, len(path))
		copy(tmp, path)
		res = append(res, tmp)
		return
	}

	for i := cutPos; i < len(s); i++ {
		sub := s[cutPos : i+1]
		if palindrome(sub) {
			path = append(path, sub)
		} else {
			continue
		}
		backTracking(s, i+1)
		path = path[:len(path)-1]
	}
}

func partition(s string) [][]string {
	res = make([][]string, 0)
	path = make([]string, 0)
	backTracking(s, 0)
	return res
}

func main() {
	res := partition("aab")
	fmt.Println(res)	
}

File: Back-tracking/131-partition/131.py
from typing import List

class Solution:
    def __init__(self):
        self.path: List[str] = []
        self.res: List[List[str]] = []

    def palindrome(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        while l <= r:
            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        return True

    def back_tracking(self, s: str, cut_pos: int):
        if cut_pos == len(s):
            self.res.append(self.path.copy())
            return
        
        for i in range(cut_pos, len(s)):
            if self.palindrome(s[cut_pos:i+1]):
                self.path.append(s[cut_pos:i+1])
            else:
                continue
            self.back_tracking(s, i + 1)
            self.path = self.path[:-1]

    def partition(self, s: str) -> List[List[str]]:
        self.back_tracking(s, 0)
        return self.res